### ISSUE TITLE
CPU multiprocessing support?

### ISSUE THREAD

**Issue by dizcza** [pwollstadt/IDTxl#35]
We have a cluster of CPU nodes where I'd like to try your package on our neurophysiological data, but it seems that IDTxl does not support multiprocessing.

**Comment by jlizier** [comment]
The recommended approach to multiprocessing is to run one job for each target node, by running a [single target analysis](https://github.com/pwollstadt/IDTxl/wiki/Single-Target-analysis) for each node rather than the default [whole network analysis](https://github.com/pwollstadt/IDTxl/wiki/Whole-Network-Analysis). You can submit one cluster job then for each target node.
Furthermore, each cluster job can be run either in single threaded mode or multithreaded if you have requested multiple cores per job. By default (if using the KSG estimator), each job will utilise as many cores as it has access to, so you need not change any settings to achieve that. If you need to set that explicitly (e.g. if you want to run jobs in a non-controlled environment where all jobs could actually ask for all of the cores, so they all end up competing) you can set the 'num_threads' option when using the JidtKraskovCMI estimator (go [here](http://pwollstadt.github.io/IDTxl/html/idtxl_estimators.html#jidt-estimators-cpu) and then find JidtKraskovCMI on the page).

@LNov I thought we already had an FAQ or something describing this, but I couldn't find it just now. Could you point to it, otherwise add something on this to the documentation and close the issue.

**Comment by mwibral** [comment]
Hi Danilo,

if you use the CPU estimators, IDTxl is recurring on the jidt 
estimators. These are multithreaded by default for the kraskov case 
(where it really matters). Given the hierachical strategy of statistical 
testing employed by IDtxl (see Novelli et al. network neuroscience, 
2019), we typically recommend to run separate analyses for each target 
(as these computations are independent under this specific statistical 
test).

Best,

Michael

**Comment by dizcza** [comment]
> You can submit one cluster job then for each target node.

How to combine the results then? I don't see this is explained in the link you specified.

Basically, I try running `demo_multivariate_te` example. How do I make it using single target analysis? BTW, when I run this example, it utilizes only one CPU core on my laptop.

**Comment by LNov** [comment]
You can combine the results using the combine_results method of the Result object (see documentation: https://pwollstadt.github.io/IDTxl/html/idtxl.html?highlight=combine_results#idtxl.results.Results.combine_results). We'll add an example in the Wiki in the future.

The demo_multivariate_te example uses one core only because it is employing the Gaussian estimator. Try using 'cmi_estimator': 'JidtKraskovCMI' instead in the settings and you should see that all the cores are used. This will be much slower of course.

On Tue, 21 May 2019 at 19:17, Danylo Ulianych <notifications@github.com<mailto:notifications@github.com>> wrote:

You can submit one cluster job then for each target node.

How to combine the results then? I don't see this is explained in the link you specified.

Basically, I try running demo_multivariate_te example. How do I make it using single target analysis? BTW, when I run this example, it utilizes only one CPU core on my laptop.

**Comment by LNov** [comment]
I updated the Wiki with an example of parallel analysis on a computing cluster using the PBS Job Scheduler:

https://github.com/pwollstadt/IDTxl/wiki/Parallel-Analysis-Using-PBS-Job-Scheduler

I also added a tutorial on how to combine the result dictionaries:

https://github.com/pwollstadt/IDTxl/wiki/Combine-Single-Target-Results
